src/Storages/Kafka/StorageKafka.cpp

```diff
@@ -473,26 +498,108 @@ KafkaConsumerPtr StorageKafka::popConsumer()
+    std::optional<size_t> closed_consumer_index;
+    for (size_t i = 0; i < consumers.size(); ++i)
+    {
+        auto & consumer_ptr = consumers[i];
+
+        if (consumer_ptr->isInUse())
+            continue;
+
+        if (consumer_ptr->hasConsumer())
+        {
+            ret_consumer_ptr = consumer_ptr;
+            break;
+        }
+
+        if (!closed_consumer_index.has_value() && !consumer_ptr->hasConsumer())
+        {
+            closed_consumer_index = i;
+        }
+    }
+
+    /// 1. There is consumer available - return it.
+    if (ret_consumer_ptr)
+    {
+        /// Noop
+    }
+    /// 2. There is no consumer, but we can create a new one.
+    else if (!ret_consumer_ptr && closed_consumer_index.has_value())
+    {
+        ret_consumer_ptr = consumers[*closed_consumer_index];
+        /// It should be OK to create consumer under lock, since it should be fast (without subscribing).
+        ret_consumer_ptr->setConsumer(createConsumer(*ret_consumer_ptr, *closed_consumer_index));
+    }
+    /// 3. There is no free consumer and num_consumers already created, waiting @timeout.
     else
     {
-        if (!semaphore.tryWait(timeout.count()))
-            return nullptr;
+        cv.wait_for(lock, timeout, [&]()
+        {
+            /// Note we are waiting only opened, free, consumers, since consumer cannot be closed right now
+            auto it = std::find_if(consumers.begin(), consumers.end(), [](const auto & ptr)
+            {
+                return !ptr->isInUse() && ptr->hasConsumer();
+            });
+            if (it != consumers.end())
+            {
+                ret_consumer_ptr = *it;
+                return true;
+            }
+            return false;
+        });
     }
 
-    // Take the first available buffer from the list
-    std::lock_guard lock(mutex);
-    auto consumer = consumers.back();
-    consumers.pop_back();
-    CurrentMetrics::add(CurrentMetrics::KafkaConsumersInUse, 1);
-    consumer->inUse();
-    return consumer;
+    if (ret_consumer_ptr)
+    {
+        CurrentMetrics::add(CurrentMetrics::KafkaConsumersInUse, 1);
+        ret_consumer_ptr->inUse();
+    }
+    return ret_consumer_ptr;
 }
 
 
-KafkaConsumerPtr StorageKafka::createConsumer(size_t consumer_number)
+KafkaConsumerPtr StorageKafka::createKafkaConsumer(size_t consumer_number)
+{
+    /// NOTE: we pass |stream_cancelled| by reference here, so the buffers should not outlive the storage.
+    auto & stream_cancelled = thread_per_consumer ? tasks[consumer_number]->stream_cancelled : tasks.back()->stream_cancelled;
+
+    KafkaConsumerPtr kafka_consumer_ptr = std::make_shared<KafkaConsumer>(
+        log,
+        getPollMaxBatchSize(),
+        getPollTimeoutMillisecond(),
+        intermediate_commit,
+        stream_cancelled,
+        topics);
+    return kafka_consumer_ptr;
+}
+
+ConsumerPtr StorageKafka::createConsumer(KafkaConsumer & kafka_consumer, size_t consumer_number)
```

Comment:
Those names (createConsumer vs createKafkaConsumer) are quite confusing. 

Maybe we can make createConsumer a method of KafkaConsumer, and just pass a consumerConfiguration inside? 

It seems like the code will become better.